[로또 - 웹, DB] 베디 미션 제출합니다.

dpudpu opened the pull request:
클린코드 나름대로 해보려고 노력했는데
스파크 자바는 처음 써봐서 어떻게 분리해야 좋을지 모르겠습니다.
부족한 부분 많은 지적 해주시면 감사하겠습니다.
resources 폴더에 테이블 생성 쿼리하고 jdbc.properties 있습니다.

dpudpu committed:
feat: DBUtils, PropertiesUil.class 구현

connection을 얻고 닫아주는 유틸리티 구현

dpudpu committed:
docs: schema.sql

dpudpu committed:
feat: RoundDao.class 라운드 관리

dpudpu committed:
fix: RoundDao.class 커넥션 close 추가

dpudpu committed:
feat: LottoDao.class 로또 추가, 조회

dpudpu committed:
refactor: Lotto.class 팩터리 메서드 접근제어자 변경

dpudpu committed:
refactor: Lotto, LottoNo 접근제어자 변경

dpudpu committed:
refactor: LottoNo.class 캐싱 시스템 롤백

dpudpu committed:
feat: ConverterToLottoNos.class 구현

dpudpu committed:
feat: WinningLottoDao.class 구현

dpudpu committed:
refactor: LottoDao convert 클래스로 추출

dpudpu committed:
refactor: WinPrize.class 총구매금액을 외부에서 받는게 아닌 직접 구하게 변경

WinPrize의 ranks로 충분히 총 구매 금액을 구할 수 있으므로 외부에서 파라미터로 받는 방식이 아닌
직접 구하는 방식으로 변경.

dpudpu committed:
feat: WinPrizeDao.class ㄱ

dpudpu committed:
refactor: 결과 포맷팅 추출

dpudpu committed:
refactor: schema.sql 변경

dpudpu committed:
fix: Lotto.class 중복된 번호 입력 가능 버그 수정

dpudpu committed:
refactor: LottoNoCollectionTest 매직넘버 상수로 변경

dpudpu committed:
style: 포맷팅

dpudpu committed:
feat: web ui 구현

dpudpu committed:
refactor: index.html 파일명 path 수정

dpudpu committed:
refactor: DBUtils.class DriverClassName jdbc.properties에 입력하게 수정

dpudpu committed:
refactor: DBUtils.class 주석 제거, 접근제어자 축소

dpudpu commented on src/main/java/lotto/dao/LottoDao.java:
```diff
@@ -0,0 +1,59 @@
+
+public class LottoDao {
+
+    //TODO 중간에 하나라도 실패하면 롤백은 어떻게?
```
질문이 있습니다.
매번 connection을 얻고 닫아주는 작업이 상당한 시간이 걸리는거 같아서 
로또를 DB에 저장할 때 하나씩이 아닌 리스트로 넘겨서 하나의 커넥션으로 리스트를 한번에 저장하게 했는데요.
이렇게 하는게 옳은 방법일까요? 
이 방법의 문제점은 중간에 오류가 날 경우 중간까지만 db에 저장될테고 이러면 롤백 해줘야할거 같아서요.
이렇게 할 경우에는 에러 발생하면 롤백 해주는 트랜잭션 설정같은거 해줘야 할까요?

woowahanCU replied:
https://tomining.tistory.com/197

insert 문을 여러형태로 넣을수 있지요.

woowahanCU replied:
하나의 커넥션으로 자원 관리로, try catch with resources 커멘트를 다른데 달아뒀어요

dpudpu replied:
Bulk Insert 사용해보려다가 Batch라는 방법도 있어서 사용해봤습니다.
OutOfMemoryError 때문에 사이즈 체크로  if문 넣어줘서  인덴트는.. 음.. stream도 사용이 안되네요 ..

dpudpu committed:
refactor: LottosGenerator.class 수동, 오토 새로 생성해서 리턴하게 변경

void 리턴 메서드에서 값을 변경하는 로직이 있는 경우 사이드이펙트가 발생 할 가능성이 생기며 해당 메서드를 테스트하기 어려울 수 있습니다.

dpudpu committed:
refactor: Rank.class 전략열거타입으로 구현

dpudpu committed:
fix: Rank.class 결과 잘 못 구하는 문제점 수정

dpudpu committed:
feat: web ui 회차별 조회 추가

dpudpu committed:
style: 포맷팅, import 제거

dpudpu committed:
test: LottoNo 캐싱 테스트 추가

dpudpu committed:
refactor: Rank.class 전략열거타입 익명클래스에서 람다로 변경

dpudpu committed:
refactor: Exception 처리 spark api 이용하게 변경

dpudpu committed:
refactor: ResultFormat.class 결과 메시지 출력 방식 변경

dpudpu committed:
refactor: WebUILottoApplication의 구현 코드들 Service로 추출

dpudpu committed:
refactor: Service 추출

dpudpu committed:
style: 포맷팅 (메소드 순서 정렬)

dpudpu committed:
refactor: start.html 안쓰는 script 제거

dpudpu committed:
fix: 잘못된 메소드명 수정

woowahanCU reviewed (commented):
처음 localhost:8080으로 접근시에 웹이 정상적으로 동작하지 않네요.
테스트를 위해서는 localhost의 mysql에 유저를 등록해야 하네요.
혹시 h2 사용에 대한 가이드를 받지는 않았었나요?

woowahanCU commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -41,6 +41,21 @@ public static Object addLottos(Request req, Response res) throws SQLException {
         return null;
     }
 
+    private static List<String> ConvertToList(final String manualLottos) {
```
camelCase

woowahanCU commented on src/main/java/lotto/domain/LottoNo.java:
```diff
@@ -5,18 +5,25 @@
 import java.util.Objects;
 
 public class LottoNo implements Comparable<LottoNo> {
+    private final static Map<Integer, LottoNo> LOTTO_NOS = new HashMap<>();
```
List<LottoNo>도 있고 Map도 있네요.

dpudpu replied:
지난번에 말씀해주셔서 중복되는 부분이 있어서 없앴다가 
저번에 변경해도 정상동작 잘 한다고 하시기도 했고 생각해보니 필요해서 다시 넣었습니다.

https://github.com/woowacourse/java-lotto/pull/3/files/1e6d330fe83b64afcc6937b9e1af790edce6bcd0#diff-0202fdc8bec5d22b14169b6288fd8845

woowahanCU replied:
ㅋㅋㅋㅋ 과거의 저도 동일한 생각을 했었군요

woowahanCU commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -1,27 +1,113 @@
+        saveLottos(userLottos, round);
+
+        int countOfManual = manualLottos.size();
+        int countOfAuto = money.getCountOfPurchase() - countOfManual;
```
Service Layer에 로직이 있네요.

dpudpu committed:
style: convertToList camelCase 이름수정

dpudpu committed:
fix: rename 누락 반영

dpudpu reviewed (commented):
> 처음 localhost:8080으로 접근시에 웹이 정상적으로 동작하지 않네요.
> 테스트를 위해서는 localhost의 mysql에 유저를 등록해야 하네요.
> 혹시 h2 사용에 대한 가이드를 받지는 않았었나요?

네 h2 사용 가이드 따로 받은적 없습니다. 
그래서 스키마하고 db접근 파일 따로 recoreces에 넣어놨는데
말씀하신 부분 알아보고 해결하겠습니다.

dpudpu committed:
refactor: Money.getCountAuto() 추가

woowahanCU commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,33 @@
+CREATE TABLE round (
```
https://docs.spring.io/spring-boot/docs/current/reference/html/howto-database-initialization.html#howto-initialize-a-database-using-spring-jdbc

여길 참조한 걸까요. 현재는 자동으로 로딩이 안될거에요. 
schema.sql로 database 초기화를 해주는건 springboot 기능이구요.
([DataSourceInitializer](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/jdbc/DataSourceInitializer.java)를 확인해봐요.)

dpudpu replied:
한번 직접 구현해봤는데 잘 한건지는 모르겠내요..

woowahanCU commented on src/main/java/lotto/service/LottoService.java:
```diff
@@ -1,27 +1,113 @@
+        saveWinPrize(round, winningLotto);
+
+        res.redirect("result");
+        return null;
```
당첨번호를 입력하지 않더라도 예외처리 없이 그냥 저장되네요

dpudpu replied:
도메인에서 구현한 유효성 검사로 예외처리하게 만들어서 서비스에서는 따로 검사를 안했습니다.
서비스에서도 검사를 해야 좋을까요?

dpudpu committed:
feat: 수동 로또 번호 입력 하지 않을 경우. 메시지 출력 추가

dpudpu committed:
fix: test 메소드명에 test누락한 부분들 추가

woowahanCU commented on src/main/java/lotto/WebUILottoApplication.java:
```diff
@@ -1,22 +1,51 @@
+            }
+            res.redirect("/error?message=" + message);
         });
     }
```
이번 미션에서는 MVC에 대해서는 따로 피드백은 안할게요.
DAO에 대해서도요. 대신 나중에 잠깐 컨셉만 따로 이야기해드릴게요.

woowahanCU commented on src/main/java/lotto/dao/LottoDao.java:
```diff
@@ -0,0 +1,59 @@
+
+public class LottoDao {
+
+    //TODO 중간에 하나라도 실패하면 롤백은 어떻게?
```
https://stackoverflow.com/questions/8066501/how-should-i-use-try-with-resources-with-jdbc
try catch with resources를 활용해봐요

dpudpu replied:
네 했습니다! 감사합니다

woowahanCU commented on src/main/java/lotto/ConsoleUILottoApplication.java:
```diff
@@ -18,13 +18,13 @@ public static void main(String[] args) {
         Money money = Money.from(inputView.inputMoney());
         CountOfManual countOfManual = CountOfManual.of(inputView.inputCountOfManual(), money.getCountOfPurchase());
```
웹에서는 이 wrapper class를 사용하지 않는군요

dpudpu replied:
웹에서 실수로 사용을 안 했네요. 웹에 반영했습니다!

woowahanCU commented on src/main/java/lotto/dao/DBUtils.java:
```diff
@@ -0,0 +1,49 @@
+package lotto.dao;
+
+import java.sql.*;
```
util class가 dao와 같이 있는것이 다소 어색하네요.
패키지 구조에 대해서도 코드를 보고 이야기하는것이 좋을 거 같습니다.

woowahanCU commented on src/main/java/lotto/service/LottoHelper.java:
```diff
@@ -0,0 +1,30 @@
+package lotto.service;
```
LottoHelper와 generate package에 있는 클래스들의 차이는 무엇일까요?
LottoService는 웹 애플리케이션만을 위한 서비스인가요? 
service package는 어떤 의미인가요?

dpudpu replied:
다시 생각해보니 확실히 다 generator 역할을 해주더라구요. 
그리고 다시 생각해보니 도메인에서 다 할 수 있는 일 같아서 도메인으로 위임했습니다.

woowahanCU commented on src/main/java/lotto/utils/ConverterToLottoNos.java:
```diff
@@ -0,0 +1,16 @@
+
+    public static List<LottoNo> convert(final String numbers) {
+        List<LottoNo> lottoNos = new LottoNosManualGenerator(numbers.substring(1, numbers.length() - 1)).generate();
+        return lottoNos;
```
따로 utils package에 구성한 의미가 있으실까요?

dpudpu replied:
1.  유틸리티 클래스
2. 여러 클래스에서 사용
이러한 이유로 유틸리티 클래스로 만들었고 utils 패키지를 만들었는데 
좋은 방법 있을까요??

woowahanCU commented on src/main/resources/schema.sql:
```diff
@@ -0,0 +1,33 @@
+  numbers VARCHAR(30) NOT NULL,
+  bonus_no INT NOT NULL,
+  round int NOT NULL,
+  PRIMARY KEY (id)
```
domain class 는 WinningLotto가 Lotto를 가지고 있는데요
테이블구조에도 반영해보는 건 어떨까요

dpudpu replied:
WinningLotto의 lotto를 lotto 테이블에 저장하고 외래키 설정으로 조인으로 가져오라는 말씀인가요?
이렇게 할 경우에 lotto 조회할 때 WinningLotto의 로또까지 같이 조회되는 문제가 생기지 않을까 해서 완전히 나눴습니다.

이렇게 한다면 컬럼을 하나 추가해서 winningLotto 인지 구분해줘야 할까요?

woowahanCU reviewed (changes requested):
아침에 말씀드린대로, 이번 미션의 목표는 (제가 생각하기에) 웹을 사용하면서도 도메인 모델의 변경을 최소화하는 경험입니다.
다른 부분들 학습하고 싶은 욕심이 있으실텐데요. 우선순위를 잘 두고 진행하시길 바래요

dpudpu committed:
refactor: Controller와 Service 분리

dpudpu committed:
refactor: RoundController non-static으로 변경

dpudpu committed:
refactor: LottoHelper.generateLottos() 삭제

dpudpu committed:
refactor: LottoHelper.generateWinPrize() -> WinningLotto.generateWinPrize() 가 생성하게 변경

dpudpu committed:
refactor: LottoHelper.generateWinningLotto() 삭제 -> WinningLotto.class 에 위임

dpudpu committed:
refactor: Encoder.class 삭제

dpudpu committed:
refactor: LottoDao.findByRound() -> findAllByRound()로 이름 변경

dpudpu committed:
feat: test용 DB runtime용 DB 분리 (리팩토링 많이 필요)

dpudpu committed:
refactor: PropertiesUtil -> DataSource로 이름 변경

dpudpu committed:
refactor: LottoDao.class  Batch Insert, try-with-resources 구현

dpudpu committed:
test: DBConnectorTest

dpudpu committed:
refactor: TableCreator 경로 변경 및 non-static 으로 변경

dpudpu committed:
refactor: RoundDao try-with-resources 구현

dpudpu committed:
feat: TestTableCreator.class 구현

dpudpu committed:
refactor: WinPrizeDao, WinningLottoDao try-with-resources 구현

dpudpu committed:
fix: RoundService.class 라운드 증가 코드 누락되어 있어서 추가.

dpudpu committed:
test: LottoDaoTest 테스트 코드 변경

dpudpu committed:
docs: schema.sql 수정

dpudpu committed:
fix: WinPrizeDao.class excuteUpdate() 누락 추가

dpudpu committed:
style: 포맷팅, import 제거

dpudpu committed:
docs: schema.sql round insert 제거

dpudpu committed:
refactor: DB h2 memory로 변경 + 앱 실행시 테이블 생성

dpudpu commented:
## 수정 목록
- mvc (controller, service 분리)
- try with resources
- DBUtils, PropertiesUtil -> DBConnector, DataSource 이름 변경 
- 테스트용 DB h2 메모리로 변경
- 새로운 환경에서 추가 작업 없이 앱 실행 바로 되게 변경 (필요한 테이블 자동 생성)
- LottoHelper.class의 generate 역할들 각 도메인으로 위임 후 삭제
- 추상화는 너무 일이 커질거 같아서 interface까지는 못하고 서비스용, 테스트용 db 다르게 실행할 수 있게까지만 했습니다.

## service, controller 등 싱글턴 
이 부분에서 가장 오래 고민했는데요. 
``` java
public class LottoController {
    public static final LottoController INSTANCE = new LottoController(....);
```
이런 싱글턴은 의존 주입이 안되고

``` java
public class LottoController {
    public static final LottoController INSTANCE = null;
    
    public static LottoController getInstance(....){
        if(INSTANCE == null){
            INSTANCE = new LottoController(...);
        }
        return INSTANCE;
    }
```
이 방식은 서버환경에서는 싱글톤이 하나만 만들어지는 것을 보장하지 못한다고 해서 다른 방법을 생각했습니다.

얼마 전에 알게 된 스프링의 싱글턴 레지스트리 기법을 사용하고 싶어서 시도해봤습니다.
모든 생성자를 public으로 해주고 앱을 실행할 때 모든 인스턴스를 생성한 다음에 
그 인스턴스만 사용하게 구현했는데요.
문제는 하나하나 다 입력 해줘야 하는 번거로움이 있어서 자동으로 생성되게 하고 싶지만 
이건 아직 영역 밖인 거 같아서 다음에 한 번 시도해보겠습니다.

```java
public static void main(String[] args) throws Exception {

        RoundDao roundDao = new RoundDao(dbConnector);
        LottoDao lottoDao = new LottoDao(dbConnector);

        RoundService roundService = new RoundService(roundDao);
        LottoService lottoService = new LottoService(lottoDao);

```

좋은 피드백 해주셔서 재미있게 고칠 수 있었습니다
감사합니다!

dpudpu committed:
refactor: Money.class Objects.requireNonNull() 추가

dpudpu committed:
refactor: Dao null 리턴 대신 Optional 구현

db 조회 후에 값이 없을 경우 null을 리턴해줬는데
Optional을 사용해보고 싶어서 변경하였지만, 지금 상황에서는 에러처리 외에는 딱히 방법이 없는거 같아서
제대로 활용은 못했지만, 학습용으로 한 번 해봄.

woowahanCU commented on build.gradle:
```diff
@@ -12,6 +12,11 @@ dependencies {
     compile('com.sparkjava:spark-core:2.9.0')
     compile('com.sparkjava:spark-template-handlebars:2.7.1')
     compile('ch.qos.logback:logback-classic:1.2.3')
+//    runtime('mysql:mysql-connector-java:8.0.16')
```
불필요한 주석은 제거해주세요.

woowahanCU commented on src/main/java/lotto/WebUILottoApplication.java:
```diff
@@ -1,22 +1,86 @@
+
+        DBConnector dbConnector = new DBConnector(DataSource.getTestInstance());
+
+        createTable(dbConnector);
```
오호라! 
그런데 프로덕션에서도 h2를 사용하네요? 🤔

woowahanCU commented on src/main/java/lotto/controller/ErrorController.java:
```diff
@@ -0,0 +1,21 @@
+public class ErrorController {
+
+    public ErrorController() {
+    }
```
어라, 기본생성자를 작성한 이유가 있을까요?

woowahanCU commented on src/main/java/lotto/config/DataSource.java:
```diff
@@ -0,0 +1,48 @@
+
+    public static DataSource getTestInstance() {
+        return LazyHolder.INSTANCE_TEST;
+    }
```
singleton pattern을 적용해줄때 기본생성자를 private으로 작성해서 객체 생성을 막아주세요.

https://blog.seotory.com/post/2016/03/java-singleton-pattern

woowahanCU commented on src/main/java/lotto/config/DBConnector.java:
```diff
@@ -0,0 +1,24 @@
+    }
+
+    public Connection getConnection() throws SQLException {
+
```
불필요한 공백 라인 제거

woowahanCU commented on src/main/java/lotto/config/TableCreator.java:
```diff
@@ -0,0 +1,52 @@
+        if (flag) {
+            return;
+        }
+        flag = true;
```
flag를 사용하기보다 singleton pattern을 적용하고 생성자에서 create메서드를 호출하는 것은 어떨까요

woowahanCU commented on src/main/java/lotto/service/WinPrizeService.java:
```diff
@@ -0,0 +1,24 @@
+
+    public WinPrize getAllByRound(final int round) throws SQLException {
+        Optional<WinPrize> winPrize = winPrizeDao.findAllByRound(round);
+        return winPrize.orElseThrow(SQLException::new);
```
Exception에 메시지를 담아주지 않아서 아래 부분에서 NPE가 발생하고 HttpStatus를 500으로 리턴합니다.
```java
message = URLEncoder.encode(exception.getMessage(), "UTF-8");
```

woowahanCU commented on src/main/java/lotto/view/OutputConsoleView.java:
```diff
@@ -26,25 +24,14 @@ public void printLottos(final List<Lotto> lottos, int countOfManual, int countOf
-    public void printRateOfProfit(final Money money, final WinPrize winPrize) {
-        double m = winPrize.getRateOfProfit(money.value());
+    public void printRateOfProfit(final WinPrize winPrize) {
+        double m = winPrize.getRateOfProfit();
```
축약어 사용은 자제해주세요.

woowahanCU commented on src/main/java/lotto/view/ResultFormat.java:
```diff
@@ -0,0 +1,29 @@
+
+    private static String message(final WinPrize winPrize, final Rank rank) {
+        if (rank == Rank.MISS) {
+            return "";
```
StringUtils.EMPTY 등 이미 있는 api를 사용해서 의미를 분명히 드러내주세요.

woowahanCU commented on src/test/java/lotto/dao/RoundDaoTest.java:
```diff
@@ -0,0 +1,34 @@
+
+    @Test
+    void getLatestTest() {
+        assertThat(-1).isNotEqualTo(roundDao.getLatest());
```
-1이 무슨 의미인지 모르겠네요

woowahanCU commented on src/main/java/lotto/dao/RoundDao.java:
```diff
@@ -0,0 +1,62 @@
+    }
+
+    public int getLatest() {
+        String sql = "SELECT * FROM round ORDER BY id DESC LIMIT 1";
```
jdbc를 활용하거나 sql mapper 등을 사용하여 쿼리를 직접 입력할 때는 *는 지양해주세요.
스키마의 변경으로 예상치 못한 부분에서 에러가 발생할 수 있습니다.

woowahanCU commented on src/main/java/lotto/service/RoundService.java:
```diff
@@ -0,0 +1,20 @@
+    }
+
+    public int getLatestRound() {
+        return roundDao.getLatest();
```
이렇게 하면 update문은 어떻게 결과 데이터를 리턴할 수 있을까요?
KeyHolder를 활용하여 해결할 수도 있어요

https://www.baeldung.com/spring-jdbc-autogenerated-keys

woowahanCU commented on src/main/java/lotto/service/WinningLottoService.java:
```diff
@@ -0,0 +1,23 @@
+        winningLottoDao.save(winningLotto, round);
+    }
+
+
```
불필요한 공백라인 제거!

woowahanCU commented on src/test/java/lotto/dao/WinPrizeDaoTest.java:
```diff
@@ -0,0 +1,45 @@
+        winPrizeDao.save(winPrize, round);
+
+        Optional<WinPrize> expected = winPrizeDao.findAllByRound(0);
+//        System.out.println(expected.orElse(new WinPrize()).getTotalPrize());
```
불필요한 주석 제거!

woowahanCU commented on src/main/java/lotto/dao/RoundDao.java:
```diff
@@ -0,0 +1,62 @@
+        }
+    }
+
+    public int getLatest() {
```
Optional을 사용할거면 다른 Dao의 시그니처도 맞춰주어야 합니다.

woowahanCU commented on src/main/java/lotto/dao/WinPrizeDao.java:
```diff
@@ -0,0 +1,67 @@
+            ps.setInt(4, winPrize.getWinCount(Rank.THIRD));
+            ps.setInt(5, winPrize.getWinCount(Rank.FOURTH));
+            ps.setInt(6, winPrize.getWinCount(Rank.FIFTH));
+            ps.setInt(7, winPrize.getWinCount(Rank.MISS));
```
저번에 잠시 보여드렸다시피 이런 부분들은 가변인자를 활용하여 해결할 수 있습니다.
```java
	private void setParameter(PreparedStatement pstmt, Object... parameter) throws SQLException {
		for (int i = 0; i < parameter.length; i++) {
			pstmt.setObject(i + 1, parameter[i]);
		}
	}
```

woowahanCU reviewed (approved):
피드백 반영해서 깔끔하게 잘 구현하셨네요. 
schema.sql 로딩한 부분은 재미있었습니다. 💯 
몇가지 피드백 남겼으니 확인해보시고 궁금한 점 있어도 DM 남기지 마세요.
레벨 1과정 진행하느라, 고생 많았어요~ 😈
